code/utilities.py

```diff
@@ -0,0 +1,62 @@
+import apache_beam as beam 
+
+def AssignKeys(row):
+  return (str(row['collection_id']) + '-' + str(row['instance_index']), row)
+
+def print_rows(row):
+  print(row)
+
+
+class Unpack(beam.DoFn):
+  def process(self, data):
+    # IMPLEMENT THIS FUNCTION
+    usage_dict = data['usage']
+    event_dict = data['event']
+    # print("----------")
+    results = {}
+    for d in (usage_dict, event_dict): 
+      results.update(d)
+    # print(results)
+    return [results]
+
+"""
+One problem: event_stream is a stream, in reality, you can only go through a stream once. You cannot loop it over several times.
+Another problem: not all events have CPU and memory limit. Some of them would have CPU/memory limit set to zero.
+The fix: before the for loop, go through the event_stream and store all events that has positive CPU and memory limit into a list. 
+Then keep your logic untouched but in the nested for loop, instead of going through event_stream, go through the cached list of events
+"""
+
+class MergeByOrder(beam.DoFn):
+  def process(self, data):
+    # IMPLEMENT THIS FUNCTION
+    _, streams = data
+    usage_stream = streams['usage'][0]
+    event_stream = streams['event'][0]
+    # print("----------")
+    picked_event = None
+    selected_events = []
+    for event in event_stream:
+      if (event['resource_request']['cpus'] > 0 and event['resource_request']['memory'] > 0):
+        selected_events.append(event)
+        
+    for usage in usage_stream:
+      for event in selected_events:
+        if (event is not None):
+          if event['time'] <= usage['start_time']:
+            picked_event = event
+      yield dict(usage=usage, event=picked_event)
+
+
+def JoinUsageAndEvent(input_usage, input_event):
```

Comment:
This is the only function that will be exported by this module. Change other classes/functions to internal only by prefix _ to their names

Reply:
understood, fixed.